File: lib/direction.ts
import type { HeadPose } from "@/lib/head-pose";

export const YAW_DEAD_ZONE_DEG = 3;
export const PITCH_DEAD_ZONE_DEG = 3;

/**
 * Screen-space look direction after calibration.
 * +x = look toward screen right, +y = look up.
 * angleDeg: atan2(y, x); 0° = right, 90° = up; null at center.
 * magnitude: 0 in the dead zone; hypot(nx, ny) - 1 outside it.
 */
export type LookDirection = {
  x: number;
  y: number;
  angleDeg: number | null;
  magnitude: number;
};

export function classifyDirection(
  pose: HeadPose | null,
  yawDeadZoneDeg = YAW_DEAD_ZONE_DEG,
  pitchDeadZoneDeg = PITCH_DEAD_ZONE_DEG,
): LookDirection | null {
  if (pose === null) {
    return null;
  }

  const yawDz = yawDeadZoneDeg > 0 ? yawDeadZoneDeg : 1;
  const pitchDz = pitchDeadZoneDeg > 0 ? pitchDeadZoneDeg : 1;

  // Normalize into screen space: +yaw is look left → -x; +pitch is look up → +y.
  const nx = -pose.yawDeg / yawDz;
  const ny = pose.pitchDeg / pitchDz;
  const r = Math.hypot(nx, ny);

  if (r <= 1) {
    return { x: 0, y: 0, angleDeg: null, magnitude: 0 };
  }

  const x = nx / r;
  const y = ny / r;
  return {
    x,
    y,
    angleDeg: (Math.atan2(y, x) * 180) / Math.PI,
    magnitude: r - 1,
  };
}

File: src/direction.py
"""Directional classifier from calibrated, smoothed yaw/pitch."""

from __future__ import annotations

from enum import Enum

from head_pose import HeadPose

YAW_DEAD_ZONE_DEG = 8.0
PITCH_DEAD_ZONE_DEG = 8.0


class Direction(Enum):
    CENTER = "CENTER"
    LEFT = "LEFT"
    RIGHT = "RIGHT"
    UP = "UP"
    DOWN = "DOWN"


def classify_direction(
    pose: HeadPose | None,
    yaw_dead_zone_deg: float = YAW_DEAD_ZONE_DEG,
    pitch_dead_zone_deg: float = PITCH_DEAD_ZONE_DEG,
) -> Direction | None:
    """Classify smoothed calibrated pose into one of five directions."""
    if pose is None:
        return None

    yaw = pose.yaw_deg
    pitch = pose.pitch_deg
    yaw_in_zone = abs(yaw) <= yaw_dead_zone_deg
    pitch_in_zone = abs(pitch) <= pitch_dead_zone_deg

    if yaw_in_zone and pitch_in_zone:
        return Direction.CENTER

    yaw_ratio = abs(yaw) / yaw_dead_zone_deg if yaw_dead_zone_deg > 0 else 0.0
    pitch_ratio = abs(pitch) / pitch_dead_zone_deg if pitch_dead_zone_deg > 0 else 0.0

    if yaw_ratio >= pitch_ratio:
        return Direction.RIGHT if yaw > 0 else Direction.LEFT
    return Direction.DOWN if pitch > 0 else Direction.UP
